README
****************************************************************************\
  $Id: README,v 1.9 2005/11/30 20:24:57 auselton Exp $
****************************************************************************
  Copyright (C) 2001-2002 The Regents of the University of California.
  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
  Written by Andrew Uselton <uselton2@llnl.gov>
  UCRL-CODE-2006-xxx.
  
  This file is part of Mib, an MPI-based parallel I/O benchamrk
  For details, see <http://www.llnl.gov/linux/mib/>.
  
  Mib is free software; you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free
  Software Foundation; either version 2 of the License, or (at your option)
  any later version.
  
  Mib is distributed in the hope that it will be useful, but WITHOUT 
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
  for more details.
  
  You should have received a copy of the GNU General Public License along
  with Mib; if not, write to the Free Software Foundation, Inc.,
  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
****************************************************************************/

			       MIB
		    MPI (POSIX) I/O BENCHMARK
			Andrew C. Uselton
		     Updated April 5, 2006

Mib is an outgrowth of Lustre (http://www.lustre.org/) benchmarking
efforts at LLNL, though its operation does not require Lustre.  It
resembles the IOR benchmark (ftp://ftp.llnl.gov/pub/siop/ior/) in that
it carries out I/O to a parallel file system from within a coordinated
parallel MPI application.  

  Mib is an MPI I/O test that has each MPI task perform I/O of a
fixed size per system call.  The tasks first write to and then read
from files named in a conventional way.  Mib assumes that the files
are on a parallel file system. The file a task reads from is
distinct from the file written to, if possible.  This defeats
caching at the clients.  Optionally, detailed timing data for
the system calls is sent to a file (one for writes and one for
reads) after the I/O is complete.
 
  Mib is intended to be as simple as possible (but no simpler).
The results are easy to interpret. The code is easy to extend.

  Mib has all the tasks stop after the same amount of time.  This
is "stonewalling" mode.  It will also stop after all tasks complete
a fixed number of system calls.  This is non-stonewalling mode.
Whichever citerion is satisfied first within a task ends I/O from
that task.  It is possible for system call limits and time limits
to be chosen so that some tasks finish due to one limit and other
tasks due to the other limit.  I don't see how that could be
desirable, but is allowed and no special note is made if it does
happen.  In order to be sure one is in stonewalling mode or
non-stornewalling mode one needs to have a good idea about the
relative progress of the fastest and slowest tasks in a test.  For
stonewalling one sets the limit on system calls above what even the
fastest task would achieve given the time limit.  Conversely, for
non-stonewalling mode set the limit on system calls below what even
the slowest would achieve in the time limit.  Note that stonewalling
mode will give the best estimate for maximum sustained throughput,
while non-stonewalling mode will more closely reflect actual "user"
experience (users of the file system not users of mib).   

Mib has two advantages over IOR.  First, it is very simple.  Following
the idea of "do only one thing, and do it well," mib avoids the
complexity and size associated with IOR.  It loads quickly.  It is
easily modified to suit specific local needs.  Second, it will
optionally produce a "profile" of the timing of every system call.
These timings are synchronized and coordinated accross all tasks.
Post-processing tools (see the "mib-tools" RPM) can manipulate
the profiles to give a graphical view of the dynamic behavior of the
file system under the load provided by mib.  These graphs have proven
valuable in isolating performance bottlenecks in Lustre.  Mib is not
intended to replace IOR but rather to compliment it.  IOR operates in
mnay more modes and with much wider applicability than mib.

Mib is characterized by the following:
o Both MPI and non-MPI operation
o POSIX system calls only
o One file per MPI task
o The test runs exactly once with the given parameters

$mib -V
mib-1.9.6-i386
$mib -h
usage: mib [b::EFIhHl:L:Mnp:PrRs:St:VW]
See the man page for "long_opts" equivalents.
    -b [<gran>]     :  Random seeks (optional granularity) before each read.
    -E              :  Show environment of test in output.
    -F              :  Mib does not normally allow I/O to any FS
                    :  but Lustre.  This overrides the "safety".
    -I              :  Show intermediate values in output.
    -h              :  Print this message and exit.
    -H              :  Show headers in output.
    -l <call_limit> :  Issue no more than this many system calls.
    -L <time_limit> :  Do not issue new system calls after this many
                    :    seconds (limits are per phase for write and
                    :    read phases).
    -M              :  Do not use MPI even if it is available.
    -n              :  Create new files if files were already present
                    :    (will always create new files if none were
                    :    present).
    -p <profiles>   :  Output system call timing profiles to 
                    :    <profiles>.write and <profiles>.read.
    -P              :  Show progress bars during testing.
    -r              :  Remove files when done.
    -R              :  Only perform the read test.
    -s <call_size>  :  Use system calls of this size (default 512k).
                    :    Numbers may use abreviations k, K, m, or M.
    -S              :  Show signon message, including date, program
                    :    name, and version.
    -t <test_dir>   :  Required. I/O transactions to and from this
                    :    directory.
    -V              :  Print the version and exit.
    -W              :  Only perform the write test
                    :    (if -R and -W are both present both tests 
                    :     will run, but that's the default anyway).

In the LLNL Linux clusters environment the SLURM
(http://www.llnl.gov/linux/slurm/) command "srun" launches MPI
applications.  A typical invokation might be 

srun -N8 -p ltest mib -t /p/ga2/lustre-test/mib/newfiles
 |    |      |     |       |
 |    |      |     |       ---target directory
 |    |      |     -----------MPI application
 |    |      -----------------partition (group of target nodes)
 |    ------------------------number of tasks
 -----------------------------SLURM MPI launcher

This example would run 8 MPI tasks on nodes taken from the "ltest"
group of compute nodes and the mib executable would open, write to,
close, open, read from, and close, a unique file for each task all in
the given directory.  The I/O would continue for the default amount of
time or the default number of transfers (whichever comes frst), and
would involve system calls using the default amount of data.  In the
following discussion assume the following shell variables are defined
to create the above command line: 

SCOM="srun -N8 -p ltest"
MCOM="mib -t /p/ga2/lustre-test/mib/newfiles"

The most common use would include a limit on the number of system
calls, a limit on time, and a system call size.  For instance if we
wanted to perform per taks no more than 4096 system calls of 512k
bytes each and to cease issuing system calls after 60 seconds, the
command would look like this:

$SCOM $MCOM -l 1024 -L 60 -s 512k

A typical result for the above command (as observed on the ALC
cluster) would look like this:


Fri Mar 17 22:50:59 2006      8    4M  4500   300	897.07     566.20

The "-H" option provides an informative header:

$SCOM $MCOM -l 1024 -L 60 -s 512k -H

           date            tasks  xfer  call  time      write       read
                                       limit limit       MB/s       MB/s
 ------------------------ ------ ----- ----- ----- ---------- ----------
 Fri Mar 17 22:50:59 2006      8    4M  4500   300     897.07      566.20

The next most common use for mib is to generate a system casll timing
profile.  The command:

$SCOM $MCOM -l 1024 -L 60 -s 512k -p ~/alc/profile

creates two files, "~/alc/profile.write" and "~/alc/profile.read".
Each of those is a table with one line per system call, and on each
line, one floating point number per task.  Each number is the
timestamp of the corresponding system call (i.e. the MPI_Wtime() at
which it returned) on the corresponding task.  The timestamps are all
relative to a "zero" at the beginning of the test coordinated accross
all tasks and adjusted for any skew in the "MPI_Wtime()" global clock
reading.  


Apendix:
 A note on configuring, compiling, and running.

One of the (possibly poor) choices I made in designing mib was to
enable it to detect if SLURM and/or MPI was available and react
accordingly.  At congigure time the "configure" script make some
guesses about the environment based on the architecture.  These are
entirely local to LLNL.  A _real_ configure script, in the auto-conf
mold, would do things right.  This one does not.  In any case an effort
is made to identify the MPI and interconnect headers and libraries,
and a similar check should set the "rpath" option to the compiler,
since the mpi and interconnect libraries are loaded dynamically.  I
have not implemented that yet, so if the /etc/ld.so.conf file does not
explicitly identify the path to the correct library directories you
will see mib running with "NO MPI" (use the -E command line option
to get confirmation about what mib thinks its environment is).  You
can work around the problem by explicitly setting:

export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:<path to lib>"

If you have access you can put the path(s) in "/etc/ld.so.conf", though
you need to remember to run "ldconfig" afterwards.  


ChangeLog
2009-05-05 Jim Garlick <garlick@llnl.gov>

	* : tag 1.9.10

	* src/sys_wrap.c : Call perror before calling FAIL on syscall error.

	* src/miberr.h : Change ASSERT/FAIL messages to use file::line fmt.
	Have ASSERT print the condition that failed.

	* src/mib.c : Use stderr for error messages, not stdout.

	* src/sys_wrap.c : Add IOMalloc() function to return page-aligned
	buffers suitable for O_DIRECT I/O.

	* src/mib.c : Allocate I/O buffers using IOMalloc(), and open with
	O_DIRECT if -d (--directio) option was used.

	* src/options.[ch], doc/mib.1 : Add -d (--directio) option.

2008-09-27 Jim Garlick <garlick@llnl.gov>

	* : tag 1.9.9

	* slurm.[ch] : Removed.

	* src/mpi_wrap.c, src/mib.c : Drop suport for -M (no mpi) option.

2008-01-31 Jim Garlick <garlick@llnl.gov>

	* : tag 1.9.8-8

	* mib.spec : Add mvapich-shmem to BuildRequires.

2007-09-30 Jim Garlick <garlick@llnl.gov>

	* : tag 1.9.8-7chaos

	* src/slurm.[ch] : Comment out all unused slurm info from struct
	to make it clear what is being used from slurm.

	* src/config.h : Remove.

	* src/mpi_wrap.c : Explicitly call MPI functions, not via dlopen.

	* src/Makefile : Compile with mib mpicc.

	* src/Makefile : Compile mib-serial with gcc and no MPI.

2007-05-05 Jim Garlick <garlick@llnl.gov>

	* : tag 1.9.8-5chaos

	* Makefile : Removed bogus quilt target.  Call mkinstalldirs 
	with -m 755 to avoid wrong permissions on /usr/lustre.

	* scripts/quilt.sh : Removed, no quilt here.

	* mib.spec : Change deprecated Copyright to License.
	Drop version information that will be overwritten by build buddah.
	Add runes to skip debug package and not strip executables.

2006-10-30 Andrew Uselton <uselton2@llnl.gov>

	* : Removed the sum.pl script which is also in the auto-test
	package.  
	
2006-10-27 Andrew Uselton <uselton2@llnl.gov>

	* : Added a discussion of library handling in the README.
	
2006-10-04 Andrew Uselton <uselton2@llnl.gov>

	* : Put default ownership of (root, root) on the files in the spec
	file for the tools package.  
	
2006-08-07 Andrew Uselton <uselton2@llnl.gov>

	* : Tag 1.9.7

	* : Modify RPM spec file to always create the
	/var/lustredata/scripts directory.  N.B.  I'd avoided this before
	because the existance of that directory might induce someone ot
	run tests and send the results there.  That could fill the root
	partition quickly if there isn't a separate /var/lustredata
	storage allocation (NFS mount).  This makes CHAOS not install
	cleanly on clusters with no Lustre.  So ite bullet and maybe
	rethink in the future.

	* : Fixed (harmless) storage leak used for communicating timing
	results in an MPI_Gather.

2006-07-13 Andrew Uselton <uselton2@llnl.gov>

	* : Tag 1.9.6

	* : Update all files with UCRL number in preparation for release.

	* : Sanity check documentation.
	
	* : Caught and fixed a bug in options handling.  Send the correct
	argc and argv to MPI_Init if you plan on consuming command line
	parameters and options.  It would segfault trying to get to
	argv[0] when sent argc = 0, which is wrong in a couple of ways. 
	
2006-06-29 Andrew Uselton <uselton2@llnl.gov>

	* : Tag 1.9.5

	* : Modify dlopen() code to be configurable, set up elan and
	infiniband  variants.

	* : Cast the "size_t count" parameter to Write() and Read() to be
	ssize_t.  This prevents incorrect execution when the
	system call returns an error condition (-1) and misinterprets it
	as a very large number.  There is also one in Snprintf.

	* : Include "-Wextra" as a CCFLAGS flag to catch future instances of
	the above.  
	
2006-04-26 Andrew Uselton <uselton2@llnl.gov>

	* : Tag 1.9.4

	* : Add lwatch.py to the project, since lmtd actually is available
	as open source.  Modify references saying otherwise in the 
	documentations.
	
2006-04-20 Andrew Uselton <uselton2@llnl.gov>

	* : Tag 1.9.3

	* : Fix really stupid, pernicious, and fundamental bug regarding use
	of opt->flags and verbosity references in macros.  They appear in
	several places, especially base_report, before opts is
	initialized.  1) split base_report from flags test and introduce
	new conditional_report.  2) do not refer to opts in macros defined
	in options.h.  Use opts in macros only locally to mib.c.
	
2006-04-3 Andrew Uselton <uselton2@llnl.gov>

	* : Tag 1.9.2

	* : Take out the node averaging.  The BGL code has entirely forked
	and node averaging is only an issue there.  It is also less reasonable
	on non-BGL clusters because the tasks do not stay coordinated across 
	a node like they do on BGL.

	* : The -V option should include the architecture, not just the signon 
	message.

	* : regularize all the documentation so it all agrees.
	
2006-03-24 Andrew Uselton <uselton2@llnl.gov>

	* : Tag 1.9.1

	* : Add the architecture to the version string

2006-03-17 Andrew Uselton <uselton2@llnl.gov>

	* : Tag 1.9

	* : Allow for operation without MPI

	* : Introduce random read option

	* : Improve the mib.sh wrapper script

	* : Make the target a required argument to avoid accidentally
	writing to NFS and causing disruption

	* : Check for the file system type of the target and refuse to use
	NFS unless a "-F" (--force) flag is set.

	* : Add a version option that prints the version string and exits.

	* : Remove the "options" file code, use command line arguments only

	* : Add a unit test "mib-test.sh" to the tools directory

	* : Fix Mib_Init, it didn't return(mib) its result

	* : When there is no SLURM or MPI we set the rank and base to the 
	host's index.  That means the read test's file name needs the 
	base added to its modulo arithmatic.  When we have arbitrary 
	subcommunicators, if ever, that would also come up.

	* : "Open" for writes should have O_CREAT, and "Open" for reads
	should not.  Doing this makes the "Open" implementation straight
	forward.  

	* : If the statfs call fails, which it will if you gave it a bogus
	target directory, then you need to bail with an error message.
	I'd left out the "exit(1)".


options.c
/*****************************************************************************\
 *  $Id: options.c,v 1.13 2006/01/09 16:26:39 auselton Exp $
 *****************************************************************************
 *  Copyright (C) 2001-2002 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Andrew Uselton <uselton2@llnl.gov>
 *  UCRL-CODE-222725
 *  
 *  This file is part of Mib, an MPI-based parallel I/O benchamrk
 *  For details, see <http://www.llnl.gov/linux/mib/>.
 *  
 *  Mib is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License (as published by the Free
 *  Software Foundation version 2, dated June 1991.
 *  
 *  Mib is distributed in the hope that it will be useful, but WITHOUT 
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
 *  for more details.
 *  
 *  You should have received a copy of the GNU General Public License along
 *  with Mib; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\*****************************************************************************/

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/vfs.h>     /* statfs */
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#define __USE_GNU       /* strnlen is a GNU extension */
#include <string.h>     /* strncpy */
#include <ctype.h>
#include <time.h>
#include <unistd.h>
#include <stdarg.h>
#include <errno.h>
#define _GNU_SOURCE
#include <getopt.h>
#include "config.h"
#include "mpi_wrap.h"
#include "mib.h"
#include "miberr.h"
#include "sys_wrap.h"
#include "options.h"
#include "slurm.h"

Options *Make_Opts();
void check_fs(char *path, int showenv, int force);
BOOL set_string(char *v, char **strp);
BOOL set_flags(char *v, int *flagp, int mask);
BOOL set_int(char *v, int *nump);
BOOL set_longlong(char *v, long long *llnump);

Options *opts;
char *opt_str = "b::EFIhHl:L:Mnp:PrRs:St:VW";

extern Mib    *mib;
extern SLURM  *slurm;
extern int     use_mpi;
extern char   *version;
extern char   *arch;

Options *
command_line(int *argcp, char **argvp[])
{
  Options *o;
  int argc = *argcp;
  char **argv = *argvp; 
  int opt;
  int idx;
  int i;
  const struct option long_options[] = 
    {
      {"random_reads", optional_argument, NULL, 'b'},
      {"show_environment", no_argument, NULL, 'E'},
      {"force", no_argument, NULL, 'F'},
      {"show_intermediate_values", no_argument, NULL, 'I'},
      {"show_headers", no_argument, NULL, 'H'},
      {"call_limit", required_argument, NULL, 'l'},
      {"time_limit", required_argument, NULL, 'L'},
      {"no_mpi", 0 , NULL, 'M'},
      {"new", 0 , NULL, 'n'},
      {"profiles", required_argument, NULL, 'p'},
      {"show_progress", no_argument, NULL, 'P'},
      {"remove", no_argument, NULL, 'r'},
      {"read_only", no_argument, NULL, 'R'},
      {"call_size", required_argument, NULL, 's'},
      {"show_signon", no_argument, NULL, 'S'},
      {"test_dir", required_argument, NULL, 't'},
      {"version", no_argument, NULL, 'V'},
      {"write_only", no_argument, NULL, 'W'},
      {"help", no_argument, NULL, 'h'},
      {0, 0, 0, 0},
    };

  o = Make_Opts();
  while( (opt = getopt_long(argc, argv, opt_str, long_options, &idx)) != -1)
    {
      switch(opt)
	{
	case 'b' : /* random_reads */
	  set_flags("true", &(o->flags), RANDOM_READS);
	  if (optarg) set_longlong(optarg, &(o->granularity));
	  break;
	case 'E' : /* show_environment */
	  set_flags("true", &(o->verbosity), SHOW_ENVIRONMENT);
	  break;
	case 'F' : /* force */
	  set_flags("true", &(o->flags), FORCE);
	  break;
	case 'I' : /* show_intermediate_values */
	  set_flags("true", &(o->verbosity), SHOW_INTERMEDIATE_VALUES);
	  break;
	case 'H' : /* show_headers */
	  set_flags("true", &(o->verbosity), SHOW_HEADERS);
	  break;
	case 'l' : /* call_limit */
	  set_int(optarg, &(o->call_limit));
	  break;
	case 'L' : /* time_limit */
	  set_int(optarg, &(o->time_limit));
	  break;
	case 'M' : /* no MPI */
	  use_mpi = FORCE_NO_MPI;
	  break;
	case 'n' : /* new */
	  set_flags("true", &(o->flags), NEW);
	  break;
	case 'p' : /* profiles */
	  if( (optarg == NULL) && ((slurm->use_SLURM == 0) || (slurm->PROCID == 0)) ) usage();
	  set_string(optarg, &(o->profiles));
	  break;
	case 'P' : /* show_progress */
	  set_flags("true", &(o->verbosity), SHOW_PROGRESS);
	  break;
	case 'r' : /* remove */
	  set_flags("true", &(o->flags), REMOVE);
	  break;
	case 'R' : /* read_only */
	  set_flags("true", &(o->flags), READ_ONLY);
	  break;
	case 's' : /* call_size */
	  set_longlong(optarg, &(o->call_size));
	  break;
	case 'S' : /* show_signon */
	  set_flags("true", &(o->verbosity), SHOW_SIGNON);
	  break;
	case 't' : /* test_dir */
	  if( (optarg == NULL) && ((slurm->use_SLURM == 0) || (slurm->PROCID == 0)) ) usage();
	  set_string(optarg, &(o->testdir));
	  break;
	case 'V' : /* version */
	  printf("mib-%s-%s\n", version, arch);
	  exit(0);
	case 'W' : /* write_only */
	  set_flags("true", &(o->flags), WRITE_ONLY);
	  break;
	case 'h' :  /* help */
	default : 
	  if ( (slurm->use_SLURM == 0) || (slurm->PROCID == 0) ) usage(); break;
	}
    }
  /* 
   * Anything we've consumed doesn't need to be passed to the MPI initializer
   */
  for(i = 1; i < *argcp - (optind - 1); i++)
      (*argvp)[i] = (*argvp)[i + (optind - 1)];
  *argcp -= (optind - 1);
  if( o->testdir == NULL)
    {
      base_report("Mib requires a \"-t <test_dir>\" argument.\n");
      exit(1);
    }
  check_fs(o->testdir, flag_set(o->verbosity, SHOW_ENVIRONMENT), flag_set(o->flags, FORCE));
  if( (o->flags & WRITE_ONLY)  && (o->flags & READ_ONLY) )
    {
      o->flags &= ~WRITE_ONLY;
      o->flags &= ~READ_ONLY;
    }
  return(o);
}

void
usage( void )
{
  printf("usage: mib [%s]\n", opt_str);
  printf("See the man page for \"long_opts\" equivalents.\n");
  printf("    -b [<gran>]     :  Random seeks (optional granularity) before each read.\n");
  printf("    -E              :  Show environment of test in output.\n");
  printf("    -F              :  Mib does not normally allow I/O to any FS\n");
  printf("                    :  but Lustre.  This overrides the \"safety\".\n");
  printf("    -I              :  Show intermediate values in output.\n");
  printf("    -h              :  Print this message and exit.\n");
  printf("    -H              :  Show headers in output.\n");
  printf("    -l <call_limit> :  Issue no more than this many system calls.\n");
  printf("    -L <time_limit> :  Do not issue new system calls after this many\n");
  printf("                    :    seconds (limits are per phase for write and\n");
  printf("                    :    read phases).\n");
  printf("    -M              :  Do not use MPI even if it is available.\n");
  printf("    -n              :  Create new files if files were already present\n");
  printf("                    :    (will always create new files if none were\n");
  printf("                    :    present).\n");
  printf("    -p <profiles>   :  Output system call timing profiles to \n");
  printf("                    :    <profiles>.write and <profiles>.read.\n");
  printf("    -P              :  Show progress bars during testing.\n");
  printf("    -r              :  Remove files when done.\n");
  printf("    -R              :  Only perform the read test.\n");
  printf("    -s <call_size>  :  Use system calls of this size (default 512k).\n");
  printf("                    :    Numbers may use abreviations k, K, m, or M.\n");
  printf("    -S              :  Show signon message, including date, program\n");
  printf("                    :    name, and version.\n");
  printf("    -t <test_dir>   :  Required. I/O transactions to and from this\n");
  printf("                    :    directory.\n");
  printf("    -V              :  Print the version and exit.\n");
  printf("    -W              :  Only perform the write test\n");
  printf("                    :    (if -R and -W are both present both tests \n");
  printf("                    :     will run, but that's the default anyway).\n");
  exit(0);
}

Options *
Make_Opts()
{
  Options *o;

  o = (Options *)Malloc(sizeof(Options));
  o->profiles = NULL;
  o->testdir = NULL;
  o->call_limit = 4096;
  o->call_size = 524288;
  o->granularity = 1;
  o->time_limit = 60;
  o->flags = DEFAULTS;
  o->verbosity = QUIET;
  return(o);
}

void
Free_Opts()
{
  if(opts->profiles != NULL) free(opts->profiles);
  if(opts->testdir != NULL) free(opts->testdir);
  if(opts != NULL)
    free(opts);
}

#define MAX_NAME 30
struct fs_es
{
  char name[MAX_NAME];
  long f_type;
  int warn;
};

const struct fs_es fs_list[] = 
  {
    {"ADFS_SUPER_MAGIC",      0xadf5, 1},
    {"AFFS_SUPER_MAGIC",      0xADFF, 1},
    {"BEFS_SUPER_MAGIC",      0x42465331, 1},
    {"BFS_MAGIC",             0x1BADFACE, 1},
    {"CIFS_MAGIC_NUMBER",     0xFF534D42, 1},
    {"CODA_SUPER_MAGIC",      0x73757245, 1},
    {"COH_SUPER_MAGIC",       0x012FF7B7, 1},
    {"CRAMFS_MAGIC",          0x28cd3d45, 1},
    {"DEVFS_SUPER_MAGIC",     0x1373, 1},
    {"EFS_SUPER_MAGIC",       0x00414A53, 1},
    {"EXT_SUPER_MAGIC",       0x137D, 1},
    {"EXT2_OLD_SUPER_MAGIC",  0xEF51, 1},
    {"EXT2_SUPER_MAGIC",      0xEF53, 1},
    {"EXT3_SUPER_MAGIC",      0xEF53, 1},
    {"HFS_SUPER_MAGIC",       0x4244, 1},
    {"HPFS_SUPER_MAGIC",      0xF995E849, 1},
    {"HUGETLBFS_MAGIC",       0x958458f6, 1},
    {"ISOFS_SUPER_MAGIC",     0x9660, 1},
    {"JFFS2_SUPER_MAGIC",     0x72b6, 1},
    {"JFS_SUPER_MAGIC",       0x3153464a, 1},
    {"MINIX_SUPER_MAGIC",     0x137F, 1},
    {"MINIX_SUPER_MAGIC2",    0x138F, 1},
    {"MINIX2_SUPER_MAGIC",    0x2468, 1},
    {"MINIX2_SUPER_MAGIC2",   0x2478, 1},
    {"MSDOS_SUPER_MAGIC",     0x4d44, 1},
    {"NCP_SUPER_MAGIC",       0x564c, 1},
    {"NFS_SUPER_MAGIC",       0x6969, 1},
    {"NTFS_SB_MAGIC",         0x5346544e, 1},
    {"OPENPROM_SUPER_MAGIC",  0x9fa1, 1},
    {"PROC_SUPER_MAGIC",      0x9fa0, 1},
    {"QNX4_SUPER_MAGIC",      0x002f, 1},
    {"REISERFS_SUPER_MAGIC",  0x52654973, 1},
    {"ROMFS_MAGIC",           0x7275, 1},
    {"SMB_SUPER_MAGIC",       0x517B, 1},
    {"SYSV2_SUPER_MAGIC",     0x012FF7B6, 1},
    {"SYSV4_SUPER_MAGIC",     0x012FF7B5, 1},
    {"TMPFS_MAGIC",           0x01021994, 1},
    {"UDF_SUPER_MAGIC",       0x15013346, 1},
    {"UFS_MAGIC",             0x00011954, 1},
    {"USBDEVICE_SUPER_MAGIC", 0x9fa2, 1},
    {"VXFS_SUPER_MAGIC",      0xa501FCF5, 1},
    {"XENIX_SUPER_MAGIC",     0x012FF7B4, 1},
    {"XFS_SUPER_MAGIC",       0x58465342, 1},
    {"_XIAFS_SUPER_MAGIC",    0x012FD16D, 1},
    {"S_MAGIC_LUSTRE",        0x0BD00BD0, 0},
    {"", 0, 0}
  };

void
check_fs(char *path, int showenv, int force)
{
  int found = 0;
  int index = 0;
  int ret;
  struct statfs buf;

  errno = 0;
  if ( (ret = statfs(path, &buf) < 0) )
  {
    base_report("Could not statfs \"%s\" (%d): %s\n", 
		path, errno, strerror(errno));
    exit(1);
  }
  while ( (!found) && (fs_list[index].f_type != 0) )
    {
      if (fs_list[index].f_type == buf.f_type) found = 1;
      else index++;
    }
  if (found)
    {
      if( (fs_list[index].warn == 1) && (!force) )
      {
	base_report("Mib will not write to a file system of type \"%s\" (f_type = %#lx) unless you use the \"-F\" (--force) option \n", 
		    fs_list[index].name, fs_list[index].f_type);
	exit(1);
      }
      if(showenv) base_report("Testing a file system of type \"%s\" (f_type = %#lx)\n",
		  fs_list[index].name, fs_list[index].f_type);
    }
  else
    {
      if(showenv) base_report("Testing an unregistered file system with f_type = %#lx\n",
		  buf.f_type);
    }
}  

void
show_details()
{
  char cluster[MAX_BUF];
  char *p;

  if ( mib->rank == mib->base )
    {
      if( slurm->use_SLURM)
	{
	  strncpy(cluster, slurm->NODELIST, MAX_BUF);
	  /* track to the end of the non-digit, non-range end of the string to get the basename */
	  for(p = cluster; ( (*p != '\0') &&
			     (*p != '[') &&
			     ( (*p < '0') || (*p > '9') ) &&
			     ( p - cluster < MAX_BUF ) ); p++);
	  /* If it's a management node just drop the 'i' */
	  if( (p > cluster) && 
	      (p - cluster < MAX_BUF) &&
	      (*(p-1) == 'i') 
	      ) 
	    *(p-1) = '\0';

	  if( (p > cluster) && 
	      (p - cluster < MAX_BUF)  && 
	      ( (*p == '[') ||
	       ( ((*p >= '0') && (*p <= '9'))) )  
	      ) 
	    *p = '\0';
	}
      else
	cluster[0] = '\0';
      printf("cluster                  = %s\n", cluster);      
      printf("new                      = %s\n", ((opts->flags & NEW) ? "true" : "false"));
      printf("remove                   = %s\n", ((opts->flags & REMOVE) ? "true" : "false"));
      printf("nodes                    = %d\n", mib->nodes);
      printf("testdir                  = %s\n", opts->testdir);
      printf("call_limit               = %d\n", opts->call_limit);
      printf("call_size                = %lld\n", opts->call_size);
      printf("time_limit               = %d\n", opts->time_limit);
      printf("tasks                    = %d\n", mib->tasks);
      printf("write_only               = %s\n", ((opts->flags & WRITE_ONLY) ? "true" : "false"));
      printf("read_only                = %s\n", ((opts->flags & READ_ONLY) ? "true" : "false"));
      printf("profiles                 = %s\n", ((opts->profiles == NULL) ? "no" : opts->profiles));
      if ( opts->flags & RANDOM_READS )
	printf("random_reads             = %lld\n", opts->granularity);
      printf("show_signon              = %s\n", ((opts->verbosity & SHOW_SIGNON) ? "true" : "false"));
      printf("show_headers             = %s\n", ((opts->verbosity & SHOW_HEADERS) ? "true" : "false"));
      printf("show_environment         = %s\n", ((opts->verbosity & SHOW_ENVIRONMENT) ? "true" : "false"));
      printf("show_progress            = %s\n", ((opts->verbosity & SHOW_PROGRESS) ? "true" : "false"));
      printf("show_intermediate_values = %s\n", ((opts->verbosity & SHOW_INTERMEDIATE_VALUES) ? "true" : "false"));
      fflush(stdout);
    }
}

BOOL
set_string(char *v, char **strp)
{
  int ret;

  if( *strp != NULL) 
      free(*strp);
  *strp = Malloc(strnlen(v, MAX_BUF) + 1);
  if ( (ret = snprintf(*strp, MAX_BUF, "%s", v)) < 0)
    FAIL();
  return(TRUE);
}

BOOL
set_flags(char *v, int *flagp, int mask)
{
  if( (v[0] == 't') ||
      (v[0] == 'y') ||
      (v[0] == '1') )
    {
      *flagp |= mask;
      return(TRUE);
    }
  if( (v[0] == 'f') ||
      (v[0] == 'n') ||
      (v[0] == '0') )
    {
      *flagp &= ~mask;
      return(TRUE);
    }
  return(FALSE);
}

BOOL
set_int(char *v, int *nump)
{
  *nump = atoi(v);
  return(TRUE);
}

BOOL
set_longlong(char *v, long long *llnump)
{
  int ret;
  char c;

  ret = sscanf(v, "%lld %c", llnump, &c);
  if (ret < 1) FAIL();
  if(ret > 1)
    {
      switch((int)c)
	{
	case 'k':
	case 'K': (*llnump) *= 1024;
	  break;
	case 'm':
	case 'M': (*llnump) *= 1024*1024;
	  break;
	default: break;
	}
    }	  
  return(TRUE);
}

